rust/src/enhancers/grammar.rs

```diff
@@ -1,179 +1,350 @@
 //! Parse enhancement rules from the string representation.
 //!
-//! The parser is made using the `nom` parser combinator library.
-//! The grammar was adapted to `nom` from:
+//! We are using a hand-written recursive descent parser. The grammar is adapted from
 //! <https://github.com/getsentry/sentry/blob/e5c5e56d176d96081ce4b25424e6ec7d3ba17cff/src/sentry/grouping/enhancer/__init__.py#L42-L79>
 
 // TODO:
-// - we should probably support better Error handling
 // - quoted identifiers/arguments should properly support escapes, etc
 
-use nom::branch::alt;
-use nom::bytes::complete::{tag, take_while1};
-use nom::character::complete::{anychar, char, space0};
-use nom::combinator::{all_consuming, map, map_res, opt, value};
-use nom::multi::{many0, many1};
-use nom::sequence::{delimited, preceded, tuple};
-use nom::{Finish, IResult, Parser};
-use smol_str::SmolStr;
+use std::borrow::Cow;
+
+use anyhow::{anyhow, Context};
 
 use super::actions::{Action, FlagAction, FlagActionType, Range, VarAction};
 use super::matchers::{FrameOffset, Matcher};
 use super::rules::Rule;
-use super::Cache;
+use super::RegexCache;
+
+/// Strips the prefix `pat` from `input` and returns the rest.
+///
+/// Returns an error if `input` doesn't start with `pat.`
+fn expect<'a>(input: &'a str, pat: &str) -> anyhow::Result<&'a str> {
+    input
+        .strip_prefix(pat)
+        .ok_or_else(|| anyhow!("at `{input}`: expected `{pat}`"))
+}
 
-fn ident(input: &str) -> IResult<&str, &str> {
-    take_while1(|c: char| c.is_ascii_alphanumeric() || matches!(c, '_' | '.' | '-'))(input)
+/// Parses a string into a bool.
+///
+/// `"1"`, `"yes"`, and `"true"` parse to `true`,
+/// `"0"`, `"no"`, and `"false"` parse to `false`,
+/// and anything else is an error.
+fn bool(input: &str) -> anyhow::Result<bool> {
+    match input {
+        "1" | "yes" | "true" => Ok(true),
+        "0" | "no" | "false" => Ok(false),
+        _ => anyhow::bail!("at `{input}`: invalid boolean value"),
+    }
 }
 
-fn rule_bool(input: &str) -> IResult<&str, bool> {
-    alt((
-        value(true, alt((tag("1"), tag("yes"), tag("true")))),
-        value(false, alt((tag("0"), tag("no"), tag("false")))),
-    ))(input)
+/// Parses an "identifier" and returns it together with the rest of the input.
+///
+/// An "identifier" is defined by the regex `[a-zA-Z0-9_.-]+`.
+fn ident(input: &str) -> anyhow::Result<(&str, &str)> {
+    let Some(end) =
+        input.find(|c: char| !(c.is_ascii_alphanumeric() || matches!(c, '_' | '.' | '-')))
```

Comment:
Depending on the length of the input, and the importance of this function, we could speed this up with a table of length 255, and make this an O(1) operation.